functions.php

```diff
@@ -0,0 +1,270 @@
+    }
+
+    return $jsonDecoded;
+}
+
+function sendResponse(string $message, int $statusCode): void
+{
+    http_response_code($statusCode);
+
+    echo $message;
+
+    exit;
+}
+
+function processURL(string &$endpoint, array &$params, string &$method): void
+{
+    $url = parse_url($_SERVER['REQUEST_URI']);
+
+    $endpoint = str_replace('/', '', $url['path']);
+
+    $params = [];
+
+    if (isset($url['query'])) {
+        parse_str($url['query'], $params);
+    }
+
+    $method = $_SERVER['REQUEST_METHOD'];
+}
+
+function readProducts(string $path): ?array
+{
+    $productsRead = @file_get_contents($path);
+
+    if ($productsRead === false) {
+        return null;
+    }
+
+    return fromJson($productsRead);
+}
+
+function extractProduct(array $products, string $category, ?int $id = null): ?array
+{
+    if (!isset($products[$category])) {
+        return null;
+    }
+
+    if ($id !== null) {
+        foreach ($products[$category] as $product) {
+            if ($product['id'] == $id) {
+                return $product;
+            }
+        }
+
+        return null;
+    }
+
+    return $products[$category];
+}
+
+function addToList(array &$products, array $requestData, ?array $oldProduct = null, bool $newId = false): void
+{
+    $fields = [ID, NAME, PRICE, QUANTITY];
+
+    foreach ($fields as $property) {
+        $updatedProduct[$property] = (
+            isset($requestData[$property]) ? $requestData[$property] :
+            (isset($oldProduct[$property]) ? $oldProduct[$property] : null)
+        );
+    }
+
+    if (!isset($products[$requestData[CATEGORY]])) {
+        $products[$requestData[CATEGORY]] = [];
+    }
+
+    if ($newId) {
+        $updatedProduct[ID] = generateId($products[$requestData[CATEGORY]], ID);
+        $products[$requestData[CATEGORY]][] = $updatedProduct;
+    } else {
+        foreach($products[$requestData[CATEGORY]] as &$product) {
+            if ($product[ID] === $updatedProduct[ID]) {
+                $product = $updatedProduct;
+            }
+        }
+    }
+}
+
+function generateId(array $categoryOfProducts, string $idKeyName): int
+{
+    $id = 0;
+
+    if (count($categoryOfProducts) > 0) {
+        foreach ($categoryOfProducts as $product) {
+            if ($product[$idKeyName] >= $id) {
+                $id = $product[$idKeyName] + 1;
+            }
+        }
+    }
+
+    return $id;
+}
+
+function validateRequestData(array $mandatoryFields, ?array $requestData = null): void
+{
+    if ($requestData === null) {
+        sendResponse(toJson(['error' => 'Sent data is not correct']), 414);
+    }
+
+    $missingFields = array_keys(array_diff_key($mandatoryFields, $requestData));
+
+    if ($missingFields) {
+        sendResponse(toJson(['error' => 'Missing fields: ' . implode(', ', $missingFields)]), 400);
+    }
+
+    $typeErrors = validateType($mandatoryFields, $requestData);
+
+    if($typeErrors) {
+        sendResponse(toJson(['error' => implode(', ', $typeErrors)]), 500);
+    }
```

Comment:
Logic of validation, should only validate, but it is even generating response with error descriptions. Logic of validation should only generate error description, but it's formatting and outputting to client should be left to the caller of this method.

Reply:
@AndreiOrmanji - Adjustments added to `validateRequestData()` function which returns the error in string format.

The caller of this function will output the errors to the client.